### PR TITLE
Create types for EmotionOverride's `theme` property

### PR DESCRIPTION
**Pull request by Brailor**

### CODE REVIEW

File: packages/types/package.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  "license": "MIT",
+  "main": "src/index.ts",
+  "types": "dist/index.d.ts",
+  "private": true,
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/instructure/instructure-ui.git"
+  },
+  "scripts": {
+    "build:types": "tsc -p tsconfig.build.json"
```
Comment (matyasf) [inline packages/types/package.json:16]: dont we need a lint script here?

File: packages/emotion/src/EmotionThemeProvider/index.tsx [hunk 2]
```diff
@@ -79,16 +104,15 @@ function EmotionThemeProvider({ children, theme = {} }: any) {
...
  * the overrides merged together.
  *
- * @param {object} theme - A full theme or an override object
+ * @param {object} themeOrOverride - A full theme or an override object
  * @returns {function} A function that returns with the theme object for the [ThemeProvider](https://emotion.sh/docs/theming#themeprovider-reactcomponenttype)
+ * @module getTheme
  */
-const getTheme = (theme: any) => (ancestorTheme = {}) => {
-  // themeable themes have a 'key' property (= name of the theme),
-  // so without it it's just an overrides objects
```
Comment (matyasf) [inline packages/emotion/src/EmotionThemeProvider/index.tsx:87]: I'd leave this comment above the `isBaseTheme` method so its clear why it works.

File: packages/types/tsconfig.build.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "extends": "../../tsconfig.build.json",
+  "compilerOptions": {
+    "outDir": "./dist"
```
Comment (matyasf) [inline packages/types/tsconfig.build.json:4]: This should output into a ./types folder, like the other packages IMO. This current solution confuses my IDE a bit (for some reason it thinks that code in ./dist is not autogenerated

File: packages/ui-avatar/src/Avatar/index.tsx [hunk 4]
```diff
@@ -80,7 +80,7 @@ category: components
 @withStyle(generateStyle, generateComponentTheme)
 @testable()
 class Avatar extends Component<Props> {
-  static componentId = 'Avatar'
+  static componentId = 'Avatar' as const
```
Comment (matyasf) [inline packages/ui-avatar/src/Avatar/index.tsx:83]: Why is this needed? Why not use `static readonly `?

File: packages/types/src/ComponentThemeMap.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+import { ThemeVariables } from './ComponentThemeVariables'
+
+type ComponentThemeVariables = {
+  Avatar: ThemeVariables['Avatar']
```
Comment (matyasf) [inline packages/types/src/ComponentThemeMap.ts:28]: as I see ThemeVariables is only used here. Instead of maintaining these 2 large arrays wouldnt it be better to just use stuff here directly from `ComponentThemeVariables`? for example: 

```
import * from './ComponentThemeVariables'

type ComponentThemeVariables = {
  Avatar: Avatar
```

File: packages/types/src/ComponentThemeVariables.ts [hunk 6]
```diff
@@ -0,0 +1,1333 @@
...
+import { Colors } from './Colors'
+
+export type Avatar = {
+  color: Colors['textBrand']
+  background: Colors['backgroundLightest']
+  borderWidthSmall: Border['widthSmall']
+  borderWidthMedium: Border['widthMedium']
+  borderColor: Colors['borderMedium']
+  fontFamily: Typography['fontFamily']
+  fontWeight: Typography['fontWeightBold']
```
Comment (ke1k0) [inline packages/types/src/ComponentThemeVariables.ts:48]: Avatar got an update since, has a few extra theme variables, please add those too. Also, this is a good example why we should type the component theme variable object too, so the IDE can warn us:

```js
// Avatar theme.js
import { Avatar } from '@instructure/types'

const generateComponentTheme = (theme) => {
    const componentVariables: Avatar = { ... }
    ...
}
```

I know it will be a pain to add it to all components, but I can help with it if you want to. I think we should do it in this PR.

File: packages/types/package.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "@instructure/types",
+  "version": "8.7.0",
+  "description": "Shared TypeScript typings for Instructure UI packages.",
+  "author": "Instructure, Inc. Engineering and Product Design",
+  "homepage": "https://instructure.design/#types",
+  "license": "MIT",
+  "main": "src/index.ts",
+  "types": "dist/index.d.ts",
+  "private": true,
```
Comment (matyasf) [inline packages/types/package.json:10]: As we've discussed this could be private, but we should state in the Readme that they should not use types that are for internal APIs because they might change.

File: packages/ui-select/src/Select/index.tsx [hunk 8]
```diff
@@ -656,25 +656,28 @@ class Select extends Component<Props> {
...
+      visibleOptionsCount,
+      children
+    } = this.props
 
     let lastWasGroup = false
     const viewProps = isShowingOptions
       ? {
-          display: 'block',
-          overflowY: 'auto',
+          display: 'block' as const,
```
Comment (ke1k0) [inline packages/ui-select/src/Select/index.tsx:669]: I'm not sure why this is needed here... I mean I can see it is needed, just don't completely understand how this works, can you please explain later? :)

File: packages/types/src/ComponentThemeVariables.ts (hunk 6 above)
Comment (ke1k0) [reply packages/types/src/ComponentThemeVariables.ts:48]: Also, because of this usage, I would rename these types to "AvatarTheme" etc, just so it indicates theme variables, not props or sth else.

File: packages/emotion/src/EmotionThemeProvider/index.tsx [hunk 9]
```diff
@@ -79,16 +104,15 @@ function EmotionThemeProvider({ children, theme = {} }: any) {
...
-  const overrides = !theme.key ? theme : null
-
-  if (!overrides) {
-    return theme
+const getTheme = (themeOrOverride: ThemeOrOverride) => (
+  ancestorTheme: BaseTheme = canvas
```
[the remark this answers is not in the document]
Comment (Brailor) [reply packages/emotion/src/EmotionThemeProvider/index.tsx:112]: I did it this way because `{}` is not assignable to the `BaseTheme` type and since we are already default to `canvas` in the `useTheme` hook, I thought this will be fine to do.

File: packages/types/src/ComponentThemeMap.ts [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+  CondensedButton: ThemeVariables['BaseButton']
+  IconButton: ThemeVariables['BaseButton']
+  Byline: ThemeVariables['Byline']
+  'Calendar.Day': ThemeVariables['CalendarDay']
+  Calendar: ThemeVariables['Calendar']
+  CheckboxFacade: ThemeVariables['CheckboxFacade']
+  ToggleFacade: ThemeVariables['ToggleFacade']
+  CodeEditor: ThemeVariables['CodeEditor']
+  'DrawerLayout.Content': ThemeVariables['DrawerLayoutContent']
+  'DrawerLayout.Tray': ThemeVariables['DrawerLayoutTray']
```
[the remark this answers is not in the document]
Comment (Brailor) [reply packages/types/src/ComponentThemeMap.ts:45]: You are right, they could be using these components directly, I only used the `componentId` property for these kind of components, but it would be better to list all of the possible variants.

File: packages/types/src/ComponentThemeVariables.ts (hunk 6 above)
Comment (Brailor) [reply packages/types/src/ComponentThemeVariables.ts:48]: Yeah, that is the reason I've made this "central" interface called `ThemeVariables` which contains all of the components theme variable typings, so one would use it like this: `ThemeVariables.Avatar` which I think is a bit less redundant than naming every component with <ComponentName>ThemeVariables.

Also I did not do the theme generator function typings on purpose in this PR, I it should be done in a different one, this PR is already pretty big.

File: packages/types/tsconfig.build.json (hunk 3 above)
Comment (Brailor) [reply packages/types/tsconfig.build.json]: Fixed

File: packages/ui-avatar/src/Avatar/index.tsx (hunk 4 above)
Comment (Brailor) [reply packages/ui-avatar/src/Avatar/index.tsx]: This was a mistake by me, I've made this change when I was exploring how to map the components to their generateTheme function, I've deleted it.

File: packages/types/src/ComponentThemeMap.ts (hunk 5 above)
Comment (Brailor) [reply packages/types/src/ComponentThemeMap.ts]: You are correct, changed

File: packages/types/package.json (hunk 7 above)
Comment (Brailor) [reply packages/types/package.json]: I've changed this to be public and will explain in the package's readme that it should be used with caution because of the internal usage.

File: packages/emotion/src/EmotionThemeProvider/index.tsx [hunk 11]
```diff
@@ -97,12 +121,19 @@ const getTheme = (theme: any) => (ancestorTheme = {}) => {
   const themeName = currentTheme.key
 
   // we pick the overrides for the current theme from the override object
-  const currentThemeOverrides = overrides?.themeOverrides?.[themeName] || {}
+  const currentThemeOverrides =
+    //@ts-expect-error fix this
```
[the remark this answers is not in the document]
Comment (Brailor) [reply packages/emotion/src/EmotionThemeProvider/index.tsx]: Done

File: packages/ui-select/src/Select/index.tsx (hunk 8 above)
Comment (Brailor) [reply packages/ui-select/src/Select/index.tsx]: I added this because of some strange TS error, but it has gone away, so this is not needed anymore, deleted.

File: packages/types/package.json (hunk 1 above)
Comment (Brailor) [reply packages/types/package.json:17]: Added the lint script

File: packages/emotion/src/EmotionThemeProvider/index.tsx (hunk 2 above)
Comment (Brailor) [reply packages/emotion/src/EmotionThemeProvider/index.tsx:87]: Done

**Review by matyasf — APPROVED** [review]

File: packages/emotion/src/EmotionThemeProvider/index.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (ke1k0) [reply packages/emotion/src/EmotionThemeProvider/index.tsx:116]: Well, not sure it is a widely used use-case, but if someone wanted to use the EmotionThemeProvider on the top level, and wrap their app in their own InstUI compatible theme, it will always merge on top of canvas then, and might cause some unwanted results.

File: packages/types/src/ComponentThemeVariables.ts (hunk 6 above)
Comment (ke1k0) [reply packages/types/src/ComponentThemeVariables.ts]: Cool, thanks! Tell me if you need help with batch-typing the generators :)

File: packages/ui-select/src/Select/index.tsx (hunk 8 above)
Comment (ke1k0) [reply packages/ui-select/src/Select/index.tsx]: Weird :D Thanks!

File: packages/emotion/src/EmotionThemeProvider/index.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (Brailor) [reply packages/emotion/src/EmotionThemeProvider/index.tsx]: You are right, changed back to be `{}` and did a type casting instead

File: packages/types/src/ComponentThemeVariables.ts (hunk 6 above)
Comment (Brailor) [reply packages/types/src/ComponentThemeVariables.ts]: I forgot to include the new Avatar theme variables, added it now

Comment (Brailor) [reply packages/types/src/ComponentThemeVariables.ts]: Also, yeah it would be nice to do the batch-typing parallel :)

**Review by ke1k0 — APPROVED** [review]
Nice job!
